### ISSUE TITLE
Nix terminates with `uncaught exception of type nix::SysError: error: getting status of /etc/ssl/certs/ca-certificates.crt` on Darwin with sandbox enabled

### ISSUE THREAD

**Issue by al3xtjames** [NixOS/nix#8485]
**Describe the bug**

After running into https://github.com/NixOS/nix/issues/7273 while running `darwin-rebuild` from [nix-darwin](https://github.com/LnL7/nix-darwin), I tried deleting `/nix/store/.links` and rebuilding. `darwin-rebuild` then failed with the following error:

```
libc++abi: terminating with uncaught exception of type nix::SysError: error: getting status of /etc/ssl/certs/ca-certificates.crt: Operation not permitted
/private/tmp/nix-build-options-db.xml.drv-0/.attr-0l2nkwhif96f51f4amnlf414lhl4rv9vh8iffyp431v6s28gsr90: line 10: 75243 Abort trap: 6           nix-instantiate --store dummy:// --eval --xml --strict --expr '{file}: builtins.fromJSON (builtins.readFile file)' --argstr file /nix/store/9pmqhvmmmjphfs0k3n51fdmxrrcb39mh-options.json > options.xml
```

**Steps To Reproduce**

1. Enable the sandbox (I set `nix.settings.sandbox = true` in my flake, which sets `sandbox = true` in /etc/nix/nix.conf).
2. Somehow reproduce https://github.com/NixOS/nix/issues/7273
3. `sudo rm -rf /nix/store/.links`
4. `darwin-rebuild --flake .#`

I'm not sure how to easily reproduce this since I only saw it after encountering https://github.com/NixOS/nix/issues/7273.

**Expected behavior**

`darwin-rebuild` should have succeeded.

**`nix-env --version` output**

```
nix-env (Nix) 2.13.3
```

**Additional context**

I see that Nix adds the cert bundle to the chroot on Linux here: https://github.com/NixOS/nix/blob/bf7dc3c7dc24f75fa623135750e8f10b8bcd94f9/src/libstore/build/local-derivation-goal.cc#L1780-L1781

On Darwin, should Nix add it to the sandbox profile? Maybe like this (untested):

```patch
diff --git a/src/libstore/build/local-derivation-goal.cc b/src/libstore/build/local-derivation-goal.cc
index b0289ac75..8b679c12f 100644
--- a/src/libstore/build/local-derivation-goal.cc
+++ b/src/libstore/build/local-derivation-goal.cc
@@ -2014,11 +2014,15 @@ void LocalDerivationGoal::runChild()
                     #include "sandbox-defaults.sb"
                     ;
 
-                if (!derivationType.isSandboxed())
+                if (!derivationType.isSandboxed()) {
                     sandboxProfile +=
                         #include "sandbox-network.sb"
                         ;
 
+                    if (settings.caFile != "")
+                        sandboxProfile += fmt("(allow file-read*\n\t(literal \"%s\")\n", settings.caFile);
+                }
+
                 /* Add the output paths we'll use at build-time to the chroot */
                 sandboxProfile += "(allow file-read* file-write* process-exec\n";
                 for (auto & [_, path] : scratchOutputs)
```

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by purcell** [comment]
Yup, I'm seeing this issue too. I have the sandbox enabled and also auto-optimise-store (as in #7273). I also tried adding `/etc/sss/certs` to `extra-sandbox-paths` via `nix-darwin`, but this made no difference.

**Comment by purcell** [comment]
Is it possible this is the same issue that was fixed by @thufschmitt in #8240? If so, I think it's in Nix 2.16, which hasn't hit `nixpkgs` yet.

**Comment by purcell** [comment]
@al3xtjames Fwiw, I seem to have worked around this issue locally by disabling `documentation.enable`, then updating `nixpkgs` (unstable), setting `nix.package = pkgs.nixVersions.nix_2_16;` and rebuilding. (All with sandbox left enabled.)

**Comment by purcell** [comment]
Yeah, there's a somewhat related issue in `home-manager` here which was fixed by bumping nmd: https://github.com/nix-community/home-manager/issues/3612

**Comment by utkarshgupta137** [comment]
I install all of my packages using `home-manager`, so `darwin-rebuild switch` is working for me while `home-manager switch` wasn't. Only thing that helped was setting [manual.manpages.enable](https://nix-community.github.io/home-manager/options.html#opt-manual.manpages.enable) to false.
My nix-config: https://github.com/utkarshgupta137/nix-config

**Comment by emilazy** [comment]
This works for me with `pkgs.nix.nixVersions.2_16` (`nix (Nix) 2.16.1`; test case: `nix build home-manager#docs-manpages`) but not with `pkgs.nix` (`nix (Nix) 2.15.1`). I have no idea what change fixed it and there's a lot of comitts to wade through. Since this breaks home-manager out of the box when using the Darwin sandbox and it'll probably be a while before 2.16 trickles down into `pkgs.nix` (never mind 23.05's entire support period), it'd be great if someone could track down what the relevant commits were and backport them to 2.15.

**Comment by emilazy** [comment]
> Is it possible this is the same issue that was fixed by @thufschmitt in #8240? If so, I think it's in Nix 2.16, which hasn't hit `nixpkgs` yet. (Or, at least, isn't the default Nix version there.)

Okay, it is just this :) I wrote this PR off because of the title and missed the additional commits that are clearly related until I searched the git log myself. Opened a backport for 2.15: https://github.com/NixOS/nix/pull/8535

**Comment by al3xtjames** [comment]
Thanks for investigating this! I was able to run `darwin-rebuild` after adding your backport:

```nix
final: prev: {
  nix = prev.nix.overrideAttrs (old: {
    patches = (old.patches or []) ++ [
      (final.fetchpatch {
        url = "https://patch-diff.githubusercontent.com/raw/NixOS/nix/pull/8535.patch";
        hash = "sha256-CTbwtlB/veZWP+xqYSSZeUJiAUhIn3r6QJgZwRx4QRM=";
      })
    ];
  });
}
```

Using Nix 2.16 also worked:

```nix
final: prev: {
  nix = final.nixVersions.nix_2_16;
}
```

**Comment by xokdvium** [comment]
Fixed in https://github.com/NixOS/nix/pull/8535 and https://github.com/NixOS/nix/pull/8240.
